### ISSUE TITLE
Improve self-hosting the Builder

### ISSUE THREAD

**Issue by johnsicili** [webstudio-is/webstudio#3966]
As a user, I want to self-host the builder and publish Projects to a platform of my choice.

**The builder is open source and can be run locally without much hassle, but isn't easy to use in a production env.**

- [ ] Allow publishing to other platforms https://github.com/webstudio-is/webstudio/issues/3654
- [ ] Todo

**Comment by kof** [comment]
This issue will also enable projects like Coolify.

**Comment by samsharpe** [comment]
Shame it's closed John, I think we got off on the wrong foot but are coming from similar perspectives. Sincere apologies if I appeared opinionated but I am. Respect.

**Comment by Silfalion** [comment]
Hi 👋 

Is it possible to self host without publishing? The export function is more than enough for my needs right now. 

I'd like to try to run it on coolify if that's possible.

**Comment by kof** [comment]
> Hi 👋
> 
> Is it possible to self host without publishing? The export function is more than enough for my needs right now.
> 
> I'd like to try to run it on coolify if that's possible.

No, publish creates a build which you are downloading with the CLI for self-hosting. Publish on a subdomain.wstd.io is not indexed by search engines.

**Comment by Silfalion** [comment]
Sorry for the late reply and thank you for clarifying. 

So it means that I can still export the websiteif I self host it on coolify and then host it(the website) wherever I want but publishing through that instance is not possible, is this correct?

**Comment by kof** [comment]
> publishing through that instance is not possible, is this correct?

Not sure what you mean

**Comment by Silfalion** [comment]
Just wanted to make sure I understood correctly that the publishing button didn't work yet on self hosted instances. All good now, downloading through the cli is all that I need, thank you for the clarification:)

**Comment by kof** [comment]
> Just wanted to make sure I understood correctly that the publishing button didn't work yet on self hosted instances. All good now, downloading through the cli is all that I need, thank you for the clarification:)

publishing can work from self-hosted instance if you set it up, you need a build pipeline, builder calls a webhook which is supposed to start build and publish process, you need to create a hook and setup the build process

We use github actions for this.

**Comment by Silfalion** [comment]
Oh I see, my assumption was that it doesn't have a build process baked in but I didn't realize it was that easy to connect to an external service.

So if I understood correctly:

- the webhook would trigger the build process in a service of your choice(GitHub actions for example)
- the script uses the cli to download the project's files
- builds it
- then publishes to the hosting provider you'd have setup in the script

**Comment by kof** [comment]
> Oh I see, my assumption was that it doesn't have a build process baked in but I didn't realize it was that easy to connect to an external service.
> 
> So if I understood correctly:
> 
> * the webhook would trigger the build process in a service of your choice(GitHub actions for example)
> * the script uses the cli to download the project's files
> * builds it
> * then publishes to the hosting provider you'd have setup in the script

exactly

**Comment by Silfalion** [comment]
That's awesome! Thank you so much for the explanation. I'm definitely gonna try to look into it soon

**Comment by vianmora** [comment]
But, @kof there is no way to contribute to this issue then ?

**Comment by kof** [comment]
> But, [@kof](https://github.com/kof) there is no way to contribute to this issue then ?

happy to discuss contributions,  discussion must happen before a PR, a PR must focus on one thing only, we won't merge things hat aren't good enough or don't fit well, feel free to also reach out on discord directly to me if you want to chat

**Comment by vianmora** [comment]
So, for the moment, i've put all the process for self-hosting in this repo : https://github.com/webstudio-community/webstudio-self-host (that is not a fork, only docker-compose) in the git-hub org associated, there's a fork with some fixes specifics to self-hosting
